proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,717 @@
+  uint32_t Flags;
+};
+
+struct ContainerDescriptorRange {
+  hlsl::RootSignature::DescriptorRangeType RangeType;
+  uint32_t NumDescriptors;
+  uint32_t BaseShaderRegister;
+  uint32_t RegisterSpace;
+  uint32_t Flags;
+  uint32_t OffsetInDescriptorsFromTableStart;
+};
+
+struct ContainerRootDescriptorTable {
+  llvm::SmallVector<RootSignature::ContainerDescriptorRange, 8>
+      DescriptorRanges;
+};
+
+struct ContainerRootConstants {
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  uint32_t Num32BitValues;
+};
+
+struct ContainerRootParameter {
+  hlsl::RootSignature::RootParameterType ParameterType;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+  std::variant<RootSignature::ContainerRootConstants,
+                   RootSignature::ContainerRootDescriptor,
+                   RootSignature::ContainerRootDescriptorTable> Extra;
+};
+
+struct StaticSamplerDesc {
+  hlsl::RootSignature::Filter Filter =
+      hlsl::RootSignature::Filter::ANISOTROPIC;
+  hlsl::RootSignature::TextureAddressMode AddressU =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  hlsl::RootSignature::TextureAddressMode AddressV =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  hlsl::RootSignature::TextureAddressMode AddressW =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  float MipLODBias = 0;
+  uint32_t MaxAnisotropy = 16;
+  hlsl::RootSignature::ComparisonFunc ComparisonFunc =
+      hlsl::RootSignature::ComparisonFunc::LessEqual;
+  hlsl::RootSignature::StaticBorderColor BorderColor =
+      hlsl::RootSignature::StaticBorderColor::OpaqueWhite;
+  float MinLOD = 0;
+  float MaxLOD = hlsl::RootSignature::Float32Max;
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+};
+
+struct ParsedRootSignature {
+  RootSignature::RootSignatureVersion Version;
+  uint32_t RootFlags;
+  llvm::SmallVector<RootSignature::ContainerRootParameter, 8>
+      RSParameters;
+  llvm::SmallVector<RootSignature::StaticSamplerDesc, 8> StaticSamplers;
+};
+
+}
+
+```
+
+This will require more work for print and serialization the AST node.
+
+The AST node serialization code might be looked like this:
+```
+void ASTDeclWriter::VisitHLSLRootSignatureDecl(HLSLRootSignatureDecl *D) {
+  VisitNamedDecl(D);
+  Record.AddSourceLocation(D->getKeywordLoc());
+  hlsl::ParsedRootSignature &RS = D->getRootSignature();
+  Record.push_back(RS.RSDesc.Version);
+  Record.push_back(RS.RSDesc.Flags);
+  for (const auto &P : ParsedRS.RSParameters) {
+    // push_back all the fields of root parameters.
+  }
+  for (const auto &SS : ParsedRS.StaticSamplers) {
+    // push_back all the fields of static samplers.
+  }
+  for (const auto &Range : ParsedRS.DescriptorRanges) {
+    // push_back all the descriptor ranges.
+  }
+  
+  Code = serialization::DECL_HLSL_ROOT_SIGNATURE;
+}
+
+```
```

Comment:
I think that this code is now out of date with the latest versions of the structs in this document.

I don't know that you need the code here.  Are you saying that the root signature will the be serialized using standard LLVM bitcode?

Reply:
The pseudocode is also not correct. When writing arrays to bitcode (as with pretty much any binary serialization) you need to push the size of the array before the `N` elements so that when you decode you know how many elements to read.

Reply:
Fixed.